--- wireless_tools.29.org/iwlib.c	2011-07-27 17:50:16.000000000 +0800
+++ wireless_tools.29/iwlib.c	2011-07-27 17:50:46.000000000 +0800
@@ -708,7 +708,7 @@
 
   /* Get ESSID */
   wrq.u.essid.pointer = (caddr_t) info->essid;
-  wrq.u.essid.length = IW_ESSID_MAX_SIZE + 1;
+  wrq.u.essid.length = IW_ESSID_MAX_SIZE;
   wrq.u.essid.flags = 0;
   if(iw_get_ext(skfd, ifname, SIOCGIWESSID, &wrq) >= 0)
     {
